src/core/server/http/router/request.ts

```diff
@@ -215,11 +216,24 @@ export class KibanaRequest<
   }
 
   private getEvents(request: Request): KibanaRequestEvents {
-    // the response is completed, or its underlying connection was terminated prematurely
-    const finish$ = fromEvent(request.raw.res, 'close').pipe(shareReplay(1), first());
+    // the response is completed
+    const finished$ = fromEvent<void>(request.raw.res, 'close').pipe(
+      filter(() => {
+        return isCompleted(request);
+      }),
+      first()
+    );
+
+    // the response's underlying connection was terminated prematurely
+    const aborted$ = fromEvent<void>(request.raw.res, 'close').pipe(
+      filter(() => {
+        return !isCompleted(request);
+      }),
+      first(),
+      takeUntil(finished$)
+    );
```

Comment:
I initially tried to use only one source `fromEvent` observable, two destination subjects, and manually emitting to these depending on `isCompleted`, but the implementation was, in the end, more complex, and more vulnerable to edge cases such as errors. Given we were already using `fromEvent` twice before these changes, I think it's acceptable anyway.

Reply:
RxJS noob question here: do we still get any benefit from having one `fromEvent` observable?

```suggestion
    const closed$ = fromEvent<void>(request.raw.res, 'close');

    const finished$ = closed$.pipe(
      filter(() => {
        return isCompleted(request);
      }),
      first()
    );

    // the response's underlying connection was terminated prematurely
    const aborted$ = closed$.pipe(
      filter(() => {
        return !isCompleted(request);
      }),
      first(),
      takeUntil(finished$)
    );
```

I'm not sure about the internal implementation of RxJS, but it looks like it'd create only one listener to the `close` event? Question is... would it work as expected?

Reply:
From the [fromEvent](https://rxjs.dev/api/index/function/fromEvent) documentation:

> Every time resulting Observable is subscribed, event handler function will be registered to event target on given event type. When that event fires, value passed as a first argument to registered function will be emitted by output Observable. When Observable is unsubscribed, function will be unregistered from event target.

I created a small [StackBlitz](https://stackblitz.com/edit/543jah?devtoolsheight=50) to confirm that 2 listeners are created / attached either way.

Reply:
> Your proposed finished$ does not have it anymore (as opposed to finish$),

That's true, but `finished$` is only internal and the public `completed$` observable subscribes to it 'instantly', and has a replay effect, so I think it was unnecessary to have it in two places.

Reply:
Also, perhaps we can reorder the operators and use `first()` first, this way there is no need to use `takeUntil()`.

```suggestion
    const closed$ = fromEvent<void>(request.raw.res, 'close').pipe(
      shareReplay(1),
      first(),
    );
    
    const finished$ = $closed.pipe(
      filter(() => {
        return isCompleted(request);
      })
    );
    
    // the response's underlying connection was terminated prematurely
    const aborted$ = $closed.pipe(
      filter(() => {
        return !isCompleted(request);
      })
    );
```

Reply:
nevermind regarding adding replay to aborted$, it was probably very late yesterday. changes in https://github.com/elastic/kibana/pull/126184/commits/19b120bfdc498ef176659665a59af09cd514cc52 and https://github.com/elastic/kibana/pull/126184/commits/df3173bd14483a76813e623ee9f369657d11744f

Reply:
> Also, perhaps we can reorder the operators 

Definitely better, yea. PR updated.